Sources/SPMBuildCore/PluginInvocation.swift

```diff
@@ -64,45 +61,220 @@ extension PluginTarget {
+                    delegate.pluginRequestedBuildOperation(subset: .init(subset), parameters: .init(parameters)) {
+                        do {
+                            switch $0 {
+                            case .success(let result):
+                                responder(try HostToPluginMessage.buildOperationResponse(result: .init(result)).toData())
+                            case .failure(let error):
+                                responder(try HostToPluginMessage.errorResponse(error: String(describing: error)).toData())
+                            }
+                        }
+                        catch {
+                            print("couldn't send reply to plugin \(error)")
+                        }
+                    }
+
+                case .testOperationRequest(let subset, let parameters):
+                    delegate.pluginRequestedTestOperation(subset: .init(subset), parameters: .init(parameters)) {
+                        do {
+                            switch $0 {
+                            case .success(let result):
+                                responder(try HostToPluginMessage.testOperationResponse(result: .init(result)).toData())
+                            case .failure(let error):
+                                responder(try HostToPluginMessage.errorResponse(error: String(describing: error)).toData())
+                            }
+                        }
+                        catch {
+                            print("couldn't send reply to plugin \(error)")
+                        }
+                    }
 
+                case .symbolGraphRequest(let targetName, let options):
+                    // The plugin requested symbol graph information for a target. We ask the delegate and then send a response.
+                    delegate.pluginRequestedSymbolGraph(forTarget: .init(targetName), options: .init(options)) {
+                        do {
+                            switch $0 {
+                            case .success(let result):
+                                responder(try HostToPluginMessage.symbolGraphResponse(result: .init(result)).toData())
+                            case .failure(let error):
+                                responder(try HostToPluginMessage.errorResponse(error: String(describing: error)).toData())
+                            }
+                        }
+                        catch {
+                            print("couldn't send reply to plugin \(error)")
+                        }
+                    }
+                }
+            }
+        }
+        let runnerDelegate = ScriptRunnerDelegate(callbackQueue: callbackQueue, delegate: delegate)
+        
         // Call the plugin script runner to actually invoke the plugin.
         scriptRunner.runPluginScript(
             sources: sources,
-            input: inputStruct,
+            initialMessage: initialMessage,
```

Comment:
Since the PluginScriptRunner no longer does the serialization, it is now passed the initial message to pass to the plugin (any message) and not necessarily one that contains a serialized package, etc. (there might not be one, depending on the kind of message it is).

Subsequent messages are then passed back to the delegate.  In the future the intent is to expand this to allow any number of messages to be sent at any time — for now having a single initial message and then letting others be follow-ons in response to requests keeps things simple and is all that's needed.